### PR TITLE
 protected erc20. version 2

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why is this change needed?

We need a canonical implementation of the Obscuro ERC20 which adds data access checks.

### What changes were made as part of this PR:

Functional PR.
- Add the right checks to the ObsErc20 
- fixes a bug where tx.origin was null for off-chain calls

### What are the key areas to look at
- ObsERC20.sol
- evm_facade

### CODE REVIEW

File: go/enclave/evm/evm_facade.go [hunk 1]
```diff
@@ -72,14 +72,18 @@ func executeTransaction(s *state.StateDB, cc *params.ChainConfig, chain *Obscuro
...
 
-	blockContext := core2.NewEVMBlockContext(convertToEthHeader(header), chain, &header.Agg)
-	vmenv := vm.NewEVM(blockContext, vm.TxContext{}, s, chainConfig, vmCfg)
+	blockContext := gethcore.NewEVMBlockContext(convertToEthHeader(header), chain, &header.Agg)
 	// todo use ToMessage
 	msg := types.NewMessage(from, &to, 0, gethcommon.Big0, 100_000, gethcommon.Big0, gethcommon.Big0, gethcommon.Big0, data, nil, true)
-	result, err := core2.ApplyMessage(vmenv, msg, gp)
+
+	// sets Tx.origin
+	txContext := gethcore.NewEVMTxContext(msg)
```
Comment (tudor-malene) [inline go/enclave/evm/evm_facade.go:83]: this is the bug fix. Before the txContexts was empty

File: integration/erc20contract/ObsERC20.sol [hunk 2]
```diff
@@ -14,12 +16,37 @@ contract ObsERC20 is ERC20 {
...
+        // 1. Human owner of an account asking for the balance.
+        // 2. Human owner of an account interacting with a smart contract which in turn asks for the balance of the original asker.
+        // Note: In case the requester spoofs the "from" of the call, they will not be able to read
+        //  the result since it will be returned encrypted with the viewing key of the declared "from".
+        if (tx.origin == account) {
+            return super.balanceOf(account);
+        }
+
+        // 3. Contract asking for its own balance.
+        if (msg.sender == account) {
```
Comment (joeldudley) [inline integration/erc20contract/ObsERC20.sol:28]: Would be clearer IMO if we combined these checks into one check, if that's possible?

Comment (tudor-malene) [reply integration/erc20contract/ObsERC20.sol:28]: it felt a bit more explicit if they're separate, with individual comments.
We'll have to think about gas cost as well.

**Review by joeldudley — APPROVED** [review]
No particular expertise here, but I can't see any mistakes.

**Pull request merged**